apex/transformer/tensor_parallel/layers.py

```diff
@@ -313,45 +320,76 @@ def forward(
 
     @staticmethod
     def backward(ctx, grad_output):
-        input, weight = ctx.saved_tensors
+        if ctx.compute_weight_gradient:
+            input, weight = ctx.saved_tensors
+        else:
+            weight = ctx.saved_tensors[0]
+            input = None
+
         use_bias = ctx.use_bias
 
-        if ctx.sequence_parallel_enabled:
-            world_size = get_tensor_model_parallel_world_size()
-            shape = list(input.shape)
-            shape[0] *= world_size
+        #only get sequence parallel inputs if need to calculate weight grad
+        handle = None
+        if ctx.compute_weight_gradient:
+            if ctx.sequence_parallel_enabled:
+                world_size = get_tensor_model_parallel_world_size()
+                shape = list(input.shape)
+                shape[0] *= world_size
+
+                all_gather_buffer = torch.empty(
+                    shape,
+                    dtype=input.dtype,
+                    device=torch.cuda.current_device(),
+                    requires_grad=False,
+                )
+                handle = torch.distributed._all_gather_base(
+                    all_gather_buffer,
+                    input,
+                    group=get_tensor_model_parallel_group(),
+                    async_op=True,
+                )
+                total_input = all_gather_buffer
+            else:
+                total_input = input
 
-            all_gather_buffer = torch.empty(
-                shape,
-                dtype=input.dtype,
-                device=torch.cuda.current_device(),
-                requires_grad=False,
-            )
-            handle = torch.distributed._all_gather_base(
-                all_gather_buffer,
-                input,
-                group=get_tensor_model_parallel_group(),
-                async_op=True,
-            )
-            total_input = all_gather_buffer
-        else:
-            total_input = input
         grad_input = grad_output.matmul(weight)
 
-        if ctx.sequence_parallel_enabled:
+        if handle is not None:
```

Comment:
Is this condition equivalent to `ctx.sequence_parallel_enabled and ctx.compute_weight_gradient`?

Reply:
Yes it is equivalent